### ISSUE TITLE
Passwords spoken aloud by screenreader on GNU/Linux

### ISSUE THREAD

**Issue by fwiep** [keepassxreboot/keepassxc#1920]
## Expected Behavior
When opening a database, or when the current database is locked, the dialog to unlock it appears. It contains, among other controls, a checkbox "Password" and a text entry for the master password. The password should not be echoed or spoken while being entered.

## Current Behavior
When using a screenreader such as Orca on Ubuntu 17.10 with GNOME 3.26.2 and typing in the master password, bullets appear visually, but the characters are spoken aloud, When adding a new entry, or editing existing entries, the passwords are read aloud, too.

## Possible Solution
Perhaps the text entries could be marked as password-field, so that the accessibility-layer recognizes this and speaks 'bullet, bullet, bullet...".

## Context
I have impaired vision and use a screenreader to interact with my computers. Being accustomed to passwords being echoed as 'bullet, bullet, bullet...", I was surprised to hear my password spoken aloud.

## Debug Info
KeePassXC - Version 2.3.1
Revision: 2fcaeea

Libraries:
- Qt 5.9.1
- libgcrypt 1.7.8

Operating system: Ubuntu 17.10
CPU architecture: x86_64
Kernel: linux 4.13.0-39-generic

Enabled extensions:
- Auto-Type
- Browser Integration
- Legacy Browser Integration (KeePassHTTP)
- SSH Agent
- YubiKey

**Comment by droidmonkey** [comment]
We will look into this asap

**Comment by fwiep** [comment]
Yes, for me it is possible to disable speech output temporarily (using Insert + S), when the password entry gets focus.

In two days I will be able to test whether the password is shown in braille, too. I guess it will be, as the braille display gets its information via the same accessibility stack.

Thanks for taking the time to fix this issue.

**Comment by droidmonkey** [comment]
At least the Braille device is rather more personal then your speakers hehe

**Comment by TheZ3ro** [comment]
I think we already flag the input as password input. Maybe it's a Qt problem.
Someone knows about a Qt program with password input to test it out?

**Comment by fwiep** [comment]
Qt-SESAM is a password manager and generation tool. See [https://github.com/ola-ct/Qt-SESAM](https://github.com/ola-ct/Qt-SESAM).

For Linux, it has to be compiled from source. I have never used it before, but do read the German magazine it is reviewed in (c't).

**Comment by phoerious** [comment]
We set the echo mode to password. The fact that this isn't passed on to assistive technologies surprises me.

**Comment by apiraino** [comment]
FWIW, I'm using:

KeePassXC - Version 2.3.2
Gnome: 3.28.1
Ubuntu: Bionic Beaver (18.04)

and I can't reproduce the issue (i.e. I launch `orca` then type in the master password in `keepassxc` and nothing is spoken aloud). If I move to a gnome-shell the screen reader starts reading.

Actually when `keepassxc` is focused the screen reader never speaks.

(just my .2 cents)

**Comment by fwiep** [comment]
Have tested the braille display: the passwords are not shown there. The screenreader does still read them aloud.

Now using an updated KeePassXC. This is my debug info:

KeePassXC - Version 2.3.3
Revision: 0a155d8

Libraries:
- Qt 5.9.1
- libgcrypt 1.7.8

Operating system: Ubuntu 17.10
CPU architecture: x86_64
Kernel: linux 4.13.0-38-generic

Enabled extensions:
- Auto-Type
- Browser Integration
- Legacy Browser Integration (KeePassHTTP)
- SSH Agent
- YubiKey

**Comment by njfox** [comment]
I tried on Arch Linux and Ubuntu 17.10--I'm not able to reproduce this. @fwiep Did you change any settings in orca to make it read the text in Qt windows? It's only reading file dialogs but no text within the actual widgets for me.

**Comment by fwiep** [comment]
Have upgraded Ubuntu, the problem still persists. This is my debug-info:

KeePassXC - Version 2.3.3
Revision: 0a155d8

Libraries:
- Qt 5.9.5
- libgcrypt 1.8.2

Operating system: Ubuntu 18.04 LTS
CPU architecture: x86_64
Kernel: linux 4.15.0-20-generic

Enabled extensions:
- Auto-Type
- Browser Integration
- Legacy Browser Integration (KeePassHTTP)
- SSH Agent
- YubiKey

**Comment by fwiep** [comment]
After another update, the problem is still present. This is my debug-info:

KeePassXC - Version 2.3.4
Revision: 6fe821c

Libraries:
- Qt 5.9.5
- libgcrypt 1.8.3

Operating system: Ubuntu 18.04.1 LTS
CPU architecture: x86_64
Kernel: linux 4.15.0-24-generic

Enabled extensions:
- Auto-Type
- Browser Integration
- Legacy Browser Integration (KeePassHTTP)
- SSH Agent
- YubiKey

**Comment by droidmonkey** [comment]
I tested this using Microsoft's Narrator on Windows 10 and it read "Asterisk-Asterisk-Asterisk...." when the password was not visible.

**Comment by droidmonkey** [comment]
I also could not reproduce on Ubuntu 19.04 using Orca.

**Comment by fwiep** [comment]
Hello, thanks for pursuing this issue further.

I can still reproduce the issue on a fresh install of Ubuntu 19.04 with KeePassXC 2.4.3 from the official PPA. I follow the these steps:

- install Ubuntu 19.04, minimal setup is sufficient
- when logged in after installation and reboot, enable the screen reader (Super + Alt + S)
- install KeePassXC from the official PPA
- start KeePassXC
- click 'create a new database'
- fill in a database name and description, click Continue
- accept the encryption settings, click Continue
- in the "Enter Password" or "Confirm Password" fields, type characters on the keyboard. They are spoken aloud by the screen reader.

This is the debug-info from Help -> About
```
KeePassXC - Version 2.4.3
Revision: 5d6ef0c

Qt 5.12.2
Debugging mode is disabled.

Operating system: Ubuntu 19.04
CPU architecture: x86_64
Kernel: linux 5.0.0-23-generic

Enabled extensions:
- Auto-Type
- Browser Integration
- SSH Agent
- KeeShare (signed and unsigned sharing)
- YubiKey

Cryptographic libraries:
 libgcrypt 1.8.4
```

**Comment by droidmonkey** [comment]
Typed characters being spoken aloud are unavoidable unless you disable it in orca settings. That doesn't have anything to do with KeePassXC, it doesn't know you are typing into a hidden password field.

What it shouldn't (and I couldn't get it to) is read the value of a password field when the password is obscured on focus/hover.

**Comment by fwiep** [comment]
You are correct! :)

I just tested having Orca's key echo disabled. The characters are no longer spoken when entered.
When I then arrow through the entered password, I hear "blank, blank, blank".
If I then press backspace and remove the characters one by one, they are yet again spoken aloud.

I don't mind this last thing - why would anyone enter his/her password completely, then remove it (thus revealing it to the listening crowd) and then type it again to actually enter it?

For me: issue solved. Thanks!
